### PR TITLE
Trying to eliminate unnecessary AsyncLocal allocation if SessionId isn't changed

### PR DESCRIPTION
**Pull request by hazzik**
This might fix the unnecessary allocations~, but I've never checked that~.

### CODE REVIEW

**Comment by hazzik** [comment]
After running a profiler it seems that this solution isn't viable.

**Comment by hazzik** [comment]
Ok, this "solution" is performing +1.53% worth than original by time.

**Comment by hazzik** [comment]
I turns out that `CallContext.LogicalSetData`/`CallContext.LogicalGetData` is much more performant than `AsyncLocal<T>`

File: src/NHibernate/Impl/SessionIdLoggingContext.cs [hunk 1]
```diff
@@ -1,18 +1,31 @@
...
 using System.Threading;
+#endif
 
 namespace NHibernate.Impl
 {
 	public class SessionIdLoggingContext : IDisposable
 	{
-		private static readonly AsyncLocal<Guid?> _currentSessionId = new AsyncLocal<Guid?>();
-
+#if NETSTANDARD2_0 || NETCOREAPP2_0
```
Comment (hazzik) [inline src/NHibernate/Impl/SessionIdLoggingContext.cs]: .net standard and .net core do not support CallContext

**Comment by hazzik** [comment]
I want this change to be introduced in a bugfix release (and release it sooner).

>Whether we accept introducing those target framework directives for this case, or if we prefer [to] avoid them as much as possible (so wait for maca solution to be released or for another one).

I introduced them beforehand as I know we're working on the .net standard version. 

> Is it still worth checking the value has changed before setting it?

Yes.. We already read the old value to store it for the future, so no extra overhead there. Then it saves us the Dispose call. So in the first version, it 1.8% of performance. Apparently _reading_ the context is most expensive operation.

> The main overhead seems to be having the context containing some data, which then gets copied at each context switch anyway.

Yeah, this overhead is still here, but it's much less than the overhead of `AsyncLocal<T>`. Someone on internet noted that `CallContext` works 5 times faster than `AsyncLocal<T>`. And it does not generate  memory traffic as opposed to `AsyncLocal<T>` (at least in _that_ scenario) - the memory consumption is below v4.1 now: 514 269 KB (526 612 072 bytes)

>If putting this solution in place, does maca's solution still provides additional significant savings? 

Yes, as above - the most expensive operation is reading the context. And we still have it here. As you've noted its 18% (here) vs 28% (when completely disabled). So we need to merge these solutions together.

**Comment by maca88** [comment]
@hazzik I did some testing and it's working fine. 👍

**Comment by hazzik** [comment]
Hi @stephencleary I’ve seen [your comment](https://stackoverflow.com/questions/37009413/callcontext-logicalgetdata-vs-threadlocal-threadstatic#comment66193557_37034175) on SO about the performance of `CallContext` vs `AsyncLocal`. It turns out that in our scenario it adds up a very significant footprint: +300MB memory allocated and a lot of CPU wasted.

**Comment by StephenCleary** [comment]
@hazzik From the [source code](https://referencesource.microsoft.com/#mscorlib/system/threading/asynclocal.cs), I just don't see how that's possible.

**Comment by hazzik** [comment]
@StephenCleary most of the CPU is wasted in dictionary operations in [ExecutionContext](https://referencesource.microsoft.com/#mscorlib/system/threading/executioncontext.cs,dcc44d583d37c571). I don't know where the allocations are coming from. But the fact is, that with `AsyncLocal<>` we got +300 MB allocated and about 25% regression by CPU.

**Comment by hazzik** [comment]
Maybe this copying of the dictionary is responsible: https://referencesource.microsoft.com/#mscorlib/system/threading/executioncontext.cs,683

**Pull request merged**